### DISCUSSION TITLE
Capture stdout of entire pipe for history

### DISCUSSION THREAD

**Discussion by byt3bl33d3r** [xonsh/xonsh#5181] [Q&A] [answered]
Heya,

I was wondering if there's a way to to capture the stdout of each command in a pipe in Xonsh? If a set `$XONSH_CAPTURE_ALWAYS = True` I seem to be only getting the stdout of the final command in the pipe by implementing my own history backend.

Take the following example:

```
echo test | awk '{print substr($1,2); }'
```

In my history backend I only see the stdout of awk:
```
{'cwd': '/workspaces/',
 'inp': "echo test | awk '{print substr($1,2); }'\n",
 'out': 'est\n',
 'rtn': 0,
 'spc': False,
 'ts': [1690664308.9095516, 1690664308.9412572]}
```

What i'm looking for is a way to see both the stdout of the echo and awk commands in this example. so ideally i'd have two history entries generated for the above command.

Is this possible?

**Comment by byt3bl33d3r** [comment]
@anki-code friendly ping. Thanks

**Reply by anki-code** [reply, marked as answer]
There is no builtin way to capture output of every command in pipe. Technically "|" is `os.pipe()`.

One way is to transform the command e.g.
```xsh
@aliases.register
def _p(args, stdin, stdout):
    """Proxy stdout."""
    for i, line in enumerate(stdin.readlines()):
        print(line, file=stdout, flush=True)
        print(f'PROXY {args} line {i}: {line}', file=__xonsh__.imp.sys.stderr)

@events.on_transform_command
def _(cmd, **kw):
    cmds = cmd.split(' /// ')
    if len(cmds) > 1:
        return ' '.join([c.strip() + f' | p {c.strip()} |' for c in cmds])[:-2]
    else:
        return cmd
```
```xsh
echo "1\n2\n3" /// grep 2  # will be `echo "1\n2\n3" | p echo "1\n2\n3" | grep 2 | p grep 2`
# PROXY ['echo', '1\n2\n3'] line 0: 1
# PROXY ['echo', '1\n2\n3'] line 1: 2
# PROXY ['echo', '1\n2\n3'] line 2: 3
# PROXY ['grep', '2'] line 0: 2
# 2
```
Another way is to patch SubprocSpec somehow. See cmds_to_specs (where "|" was implemented) and maybe DecoratorAlias (for prepare patch) in specs.py.

See also [xontrib-pipeliner](https://github.com/anki-code/xontrib-pipeliner).
